### ISSUE TITLE
"publish" attribute should be removed from entities which are not packages

### ISSUE THREAD

**Issue by iainduncani** [openwhisk/openwhisk#465]
The [docs on actions](https://github.com/openwhisk/openwhisk/blob/master/docs/actions.md) do not include any information about the --shared option
The cli just says that it is a boolean:
`usage: wsk-script.py action create [-h] [-u AUTH] [--docker] [--copy]
                                   [--sequence] [--lib LIB]
                                   **[--shared [{yes,no}]]**
                                   [-a ANNOTATION ANNOTATION] [-p PARAM PARAM]
                                   [-t TIMEOUT] [-m MEMORY]
                                   name artifact`
It is not clear what you would use this for.

**Comment by rabbah** [comment]
You're right to point this out. It exists on triggers and rules as well but is currently not used/enforced. The only effect this flag has, when applied to packages is to share or hide the contents of the package, entirely.

**Comment by bjustin-ibm** [comment]
After some experimentation, it became clear that the publish attribute for actions is not needed and, worse, can lead to confusion.

For example, create a package with --shared yes then create an action in that package also with --shared yes. Running wsk package get shows that the action has a publish value of false. However, running wsk action get on the actions shows the publish value set to true.

Additionally, running wsk action create --shared no to create an action inside a published package does not prevent others from invoking the action - that is, the publish value has no effect.

Lastly... creating an action outside of a shared package with --shared yes does not allow others to invoke that action.

tl;dr The publish attribute for actions (and [other non-package entities](https://github.com/openwhisk/openwhisk/issues/465#issuecomment-220458240) as well) doesn't do anything, and should be removed.

This will affect the backend, CLI, and documentation.

**Comment by rabbah** [comment]
Noting here:
1. the publish attribute is a common property on all entities (includes rules, triggers, actions and packages)
2. only has an effect on packages (and the visibility of the package dictates the visibility of all assets in it)
3. there was some thinking to allow more fine grained sharing (private actions in a package) but since that's not implemented, the cleanest solution is to remove the property from the base type `WhiskEntity` and CLI commands.

**Comment by dubeejw** [comment]
The shared functionality will be removed from the CLI for all commands except the package command in PR https://github.com/openwhisk/openwhisk/pull/1499.

**Comment by rabbah** [comment]
While the flag is removed from the CLI, some more work is needed on the backend to finish this.

**Comment by larandersson** [comment]
@rabbah The *publish* attribute is a required one according the spec, so removing it would break the API. Maybe we should create an issue for updating the API to v2 to address this, and several other issues? That would allow us to get all the required changes to the API in one place and close (?) the issues that require a new version of the API . We could also upgrade to OpenAPI v 3.0 at the same time.
